src/EditorFeatures/Core/Diagnostics/AbstractPushOrPullDiagnosticsTaggerProvider.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+internal static class DiagnosticTaggingOptions
+{
+    public static readonly Option2<bool> PullDiagnosticTagging = new(
+        "DiagnosticTaggingOptions", "PullDiagnosticTagging", defaultValue: true,
+        new FeatureFlagStorageLocation("Roslyn.PullDiagnosticTagging"));
+}
+
+/// <summary>
+/// Base type of all diagnostic taggers (classification, squiggles, suggestions, inline-diags).  Subclasses can control
+/// things by overriding functionality in this type.  Internally, this will switch to either a pull or push cased
+/// approach at instantiation time depending on our internal feature flag.
+/// </summary>
+internal abstract partial class AbstractPushOrPullDiagnosticsTaggerProvider<TTag> : ITaggerProvider
+    where TTag : ITag
+{
+    private readonly ITaggerProvider _underlyingTaggerProvider;
+
+    protected readonly IGlobalOptionService GlobalOptions;
+
+    protected AbstractPushOrPullDiagnosticsTaggerProvider(
+        IThreadingContext threadingContext,
+        IDiagnosticService diagnosticService,
+        IDiagnosticAnalyzerService analyzerService,
+        IGlobalOptionService globalOptions,
+        ITextBufferVisibilityTracker? visibilityTracker,
+        IAsynchronousOperationListener listener)
+    {
+        GlobalOptions = globalOptions;
+
+        if (globalOptions.GetOption(DiagnosticTaggingOptions.PullDiagnosticTagging))
```

Comment:
I'm not sure about this - we now have two different feature flags for diagnostics (4 different states for diagnostics?).  Pull/Push tagging and LSP on/off.

If pull tagging isn't stable should we leave the tagger always on push and use LSP only for pull?

Reply:
i believe pull-tagging to be stable, and strictly better than push tagging.  However, we had a perf-issue with it due to the same issue we ran into with pull tagging in LSP (namely, that we need to separate out syntax vs semantics).  The virtue of going this route is that we can independently hammer 'pull diagnostics' in general (i.e. hammering the IDiagAnalyzerService API) without having to take a dependency on LSP.  And anything we find/fix there benefits LSP as well.